ffi/src/session/mod.rs

```diff
@@ -66,6 +67,27 @@ pub mod ffi {
                 .process_fastpath_input(&mut image.0, &fastpath_input.0)
                 .map(|outputs| Box::new(ActiveStageOutputIterator(outputs)))?)
         }
+
+        pub fn get_svc_processor_cliprdr<'a>(
+            &'a mut self,
+        ) -> Result<Option<Box<CliprdrReference<'a>>>, Box<IronRdpError>> {
+            let cliprdr_ref = self.0.get_svc_processor::<ironrdp::cliprdr::CliprdrClient>();
+            cliprdr_ref
+                .map(|cliprdr| Box::new(CliprdrReference(cliprdr)))
+                .map(Some)
+                .map_or(Ok(None), Ok)
+        }
```

Comment:
issue: Unfortunately, the `CliprdrReference` abstraction is problematic for other reasons… I suggest its removal. We have very high risks of violating the aliasing model of Rust: we have a shared reference and at the same time it’s possible to mutate the ActiveStage structure with one of the `&mut self` functions.

I suggest to remove both `get_svc_processor_cliprdr` and `process_svc_processor_message_cliprdr`. Remove all the related boilerplate. Add `initiate_clipboard_copy`, `initiate_clipboard_paste` and `submit_clipboard_format_data` directly on `ActiveStage`. Process the message immediately and make the functions return `VecU8` directly.

This will be much safer to use, and requires less code.

Reply:
Will update 👍

Reply:
updated